### PR TITLE
refactor(connector): [adyen] add more fields in the payments request

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- Add shopper statement and metadata in payments request
- Add additional fields and validations over fields like `IP address`, `billing_address`, `Email` and Shopper details

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [X] This PR modifies application configuration/environment variables

Production URL got updated for Adyen

## Motivation and Context

#3914 

## How did you test it?

- Create Klarna Payment
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_6LLuoNT8iBk4hTwVLSQ6zTJOiBmeuLSJ5NoakZTZItC4IifnHLy7cTWPZXW4uhbq' \
--data-raw '{
    "amount": 6540,
    "currency": "EUR",
    "confirm": true,
    "capture_method": "manual",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://duck.com",
    "profile_id": "pro_B0AWG00VLGTFt4rTrj6a",
    "business_country": "US",
    "payment_method": "pay_later",
    "payment_method_type": "klarna",
    "payment_experience": "redirect_to_url",
    "connector": [
        "adyen"
    ],
    "payment_method_data": {
        "pay_later": {
            "klarna_redirect": {
                "billing_email": "hyperswitch_sdk_demo_id@gmail.com",
                "billing_country": "DE"
            }
        }
    },
    "routing": {
        "type": "single",
        "data": "adyen"
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "DE",
            "first_name": "PiX"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
- Test in Adyen dashboard all the fields should be visible like shopper statement, address and shopper reference
- Klarna Validation when not sending Email
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_6LLuoNT8iBk4hTwVLSQ6zTJOiBmeuLSJ5NoakZTZItC4IifnHLy7cTWPZXW4uhbq' \
--data-raw '{
    "amount": 6540,
    "currency": "EUR",
    "confirm": true,
    "capture_method": "manual",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://duck.com",
    "profile_id": "pro_B0AWG00VLGTFt4rTrj6a",
    "business_country": "US",
    "payment_method": "pay_later",
    "payment_method_type": "klarna",
    "payment_experience": "redirect_to_url",
    "connector": [
        "adyen"
    ],
    "payment_method_data": {
        "pay_later": {
            "klarna_redirect": {
                "billing_email": "hyperswitch_sdk_demo_id@gmail.com",
                "billing_country": "DE"
            }
        }
    },
    "routing": {
        "type": "single",
        "data": "adyen"
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "DE",
            "first_name": "PiX"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
Expected Response: 

```
{
    "error": {
        "type": "invalid_request",
        "message": "Missing required param: email",
        "code": "IR_04"
    }
}
```

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by swangi-kumari — DISMISSED** [review]

File: crates/router/src/connector/adyen/transformers.rs [hunk 1]
```diff
@@ -2515,12 +2593,22 @@ impl<'a>
         let amount = get_amount_data(item);
         let auth_type = AdyenAuthType::try_from(&item.router_data.connector_auth_type)?;
         let shopper_interaction = AdyenShopperInteraction::from(item.router_data);
-        let (recurring_processing_model, store_payment_method, shopper_reference) =
+        let shopper_reference = item
+            .router_data
+            .customer_id
+            .clone()
+            .map(|c_id| format!("{}_{}", item.router_data.merchant_id, c_id));
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:2600]: Move this logic inside a method and reuse

File: crates/router/src/connector/adyen/transformers.rs [hunk 2]
```diff
@@ -2076,27 +2079,81 @@ impl<'a> TryFrom<&api::WalletData> for AdyenPaymentMethod<'a> {
...
                 Ok(AdyenPaymentMethod::AdyenKlarna(Box::new(klarna)))
             }
-            api_models::payments::PayLaterData::AffirmRedirect { .. } => Ok(
-                AdyenPaymentMethod::AdyenAffirm(Box::new(PmdForPaymentType {
-                    payment_type: PaymentType::Affirm,
-                })),
-            ),
+            api_models::payments::PayLaterData::AffirmRedirect { .. } => {
+                check_required_field(shopper_email, "email")?;
+                check_required_field(shopper_name, "shopper_name")?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:2141]: Please use the same name and complete path from our payment request

File: crates/router/src/connector/adyen/transformers.rs [hunk 3]
```diff
@@ -2114,17 +2171,36 @@ impl<'a> TryFrom<(&api::PayLaterData, Option<api_enums::CountryAlpha2>)>
                 }
             }
             api_models::payments::PayLaterData::PayBrightRedirect { .. } => {
+                check_required_field(shopper_name, "shopper_name")?;
+                check_required_field(telephone_number, "telephone_number")?;
+                check_required_field(shopper_email, "email")?;
+                check_required_field(billing_address, "billing_address")?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:2177]: Shouldn't we need to validate the inner params of shipping_address and billing_address?

Comment (SamraatBansal) [reply crates/router/src/connector/adyen/transformers.rs:2177]: The inner fields requirements are not provided in the docs, the object is expected from Adyen so we are validating the object only.

Comment (SamraatBansal) [reply crates/router/src/connector/adyen/transformers.rs:2177]: Adyen docs specifies either the complete Address Object is optional, if you are passing the objects then the certain fields inside the object are required so accordingly made that change

**Review by swangi-kumari — APPROVED** [review]

**Pull request merged**